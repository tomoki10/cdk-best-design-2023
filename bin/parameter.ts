import { Environment } from "aws-cdk-lib";

export interface AppParameter {
  env?: Environment;
  envName: string;
}
export const devParameter: AppParameter = {
  envName: "Development",
  env: { account: "123456789012", region: "ap-northeast-1" },
};

export const prdParameter: AppParameter = {
  envName: "Production",
  env: { account: "012345678901", region: "ap-northeast-1" },
};
