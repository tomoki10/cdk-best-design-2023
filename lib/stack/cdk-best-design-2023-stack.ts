import * as cdk from "aws-cdk-lib";
import { Construct } from "constructs";
import { LambdaConstruct } from "../construct/lambda";
import { DataStoreConstruct } from "../construct/data-store";

interface CdkBestDesign2023StackProps extends cdk.StackProps {
  envName: string;
}

export class CdkBestDesign2023Stack extends cdk.Stack {
  constructor(
    scope: Construct,
    id: string,
    props: CdkBestDesign2023StackProps
  ) {
    super(scope, id, props);

    const dataStore = new DataStoreConstruct(this, "DataStoreConstruct", {
      envName: props.envName,
    });

    new LambdaConstruct(this, "LambdaConstruct", {
      envName: props.envName,
      tableName: dataStore.itemTableName,
    });
  }
}
