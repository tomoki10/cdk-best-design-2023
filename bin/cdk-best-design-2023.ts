import * as cdk from "aws-cdk-lib";
import {
  CdkBestDesign2023Stack,
  // CdkBestDesign2023Stack as DevCdkBestDesign2023Stack,
  // CdkBestDesign2023Stack as PrdCdkBestDesign2023Stack,
} from "../lib/stack/cdk-best-design-2023-stack";

import { devParameter, prdParameter, AppParameter } from "./parameter";
const app = new cdk.App();

// Dynamic ID Pattern
const argContext = "environment";
const envKey = app.node.tryGetContext(argContext);
const parameters = [devParameter, prdParameter];
const appParameter: AppParameter = parameters.filter(
  (obj: AppParameter) => obj.envName === envKey
)[0];

new CdkBestDesign2023Stack(
  app,
  `${appParameter.envName}CdkBestDesign2023Stack`,
  {
    ...appParameter,
  }
);

// Static ID Pattern 1:
// new CdkBestDesign2023Stack(app, `DevCdkBestDesign2023Stack`, {
//   ...devParameter,
// });
// new CdkBestDesign2023Stack(app, `PrdCdkBestDesign2023Stack`, {
//   ...prdParameter,
// });

// // Static ID Pattern 2:
// new DevCdkBestDesign2023Stack(app, `DevCdkBestDesign2023Stack`, {
//   ...devParameter,
// });

// new PrdCdkBestDesign2023Stack(app, `PrdCdkBestDesign2023Stack`, {
//   ...prdParameter,
// });
