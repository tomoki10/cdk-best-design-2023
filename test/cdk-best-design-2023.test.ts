import * as cdk from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import * as CdkBestDesign2023 from "../lib/stack/cdk-best-design-2023-stack";

import { devParameter } from "../bin/parameter";

test("Snapshot Test", () => {
  const app = new cdk.App();
  const stack = new CdkBestDesign2023.CdkBestDesign2023Stack(
    app,
    "CdkBestDesign2023Stack",
    {
      ...devParameter,
    }
  );
  expect(Template.fromStack(stack)).toMatchSnapshot();
});
