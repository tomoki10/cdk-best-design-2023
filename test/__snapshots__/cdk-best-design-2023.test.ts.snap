// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "DataStoreConstructItemTableAF46E9D4": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "id",
            "AttributeType": "S",
          },
          {
            "AttributeName": "name",
            "AttributeType": "S",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "ContributorInsightsSpecification": {
          "Enabled": true,
        },
        "DeletionProtectionEnabled": true,
        "KeySchema": [
          {
            "AttributeName": "id",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "name",
            "KeyType": "RANGE",
          },
        ],
        "PointInTimeRecoverySpecification": {
          "PointInTimeRecoveryEnabled": true,
        },
        "SSESpecification": {
          "SSEEnabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "LambdaConstructgetItem3754D7A4": {
      "DependsOn": [
        "LambdaConstructgetItemServiceRoleDefaultPolicyA66A40CD",
        "LambdaConstructgetItemServiceRoleCE9D34BB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-729929860662-ap-northeast-1",
          "S3Key": "87ebf52a9d89d33ff0abe149330d745acbea71479a8f583429e80aa3beedb246.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DDB_TABLE": {
              "Ref": "DataStoreConstructItemTableAF46E9D4",
            },
          },
        },
        "Handler": "index.getItem",
        "Layers": [
          "arn:aws:lambda:ap-northeast-1:580247275435:layer:LambdaInsightsExtension:14",
        ],
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "LambdaConstructgetItemServiceRoleCE9D34BB",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaConstructgetItemLogRetention59BD2926": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LambdaConstructgetItem3754D7A4",
              },
            ],
          ],
        },
        "RetentionInDays": 90,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaConstructgetItemServiceRoleCE9D34BB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaConstructgetItemServiceRoleDefaultPolicyA66A40CD": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaConstructgetItemServiceRoleDefaultPolicyA66A40CD",
        "Roles": [
          {
            "Ref": "LambdaConstructgetItemServiceRoleCE9D34BB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaConstructgetUserD01056E7": {
      "DependsOn": [
        "LambdaConstructgetUserServiceRole699FA980",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-729929860662-ap-northeast-1",
          "S3Key": "ded7091e9742825848513aeb3d7efa3bf2cb1c486e265c50ae6d382a38ad3241.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.getUser",
        "Role": {
          "Fn::GetAtt": [
            "LambdaConstructgetUserServiceRole699FA980",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaConstructgetUserServiceRole699FA980": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaConstructpostItem9C3F458B": {
      "DependsOn": [
        "LambdaConstructpostItemServiceRoleDefaultPolicy12F88511",
        "LambdaConstructpostItemServiceRoleE6F7D38F",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-729929860662-ap-northeast-1",
          "S3Key": "d70d38c8c5b6efad0a57f47012ece60ddcdfca00204b566a2e58075e6a42b556.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DDB_TABLE": "hoge-table",
          },
        },
        "Handler": "index.postItem",
        "Layers": [
          "arn:aws:lambda:ap-northeast-1:580247275435:layer:LambdaInsightsExtension:14",
        ],
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "LambdaConstructpostItemServiceRoleE6F7D38F",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaConstructpostItemLogRetention1759A719": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LambdaConstructpostItem9C3F458B",
              },
            ],
          ],
        },
        "RetentionInDays": 90,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaConstructpostItemServiceRoleDefaultPolicy12F88511": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaConstructpostItemServiceRoleDefaultPolicy12F88511",
        "Roles": [
          {
            "Ref": "LambdaConstructpostItemServiceRoleE6F7D38F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LambdaConstructpostItemServiceRoleE6F7D38F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaConstructputItem09D64D38": {
      "DependsOn": [
        "LambdaConstructputItemServiceRoleDefaultPolicy730237CB",
        "LambdaConstructputItemServiceRole1DA96156",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-729929860662-ap-northeast-1",
          "S3Key": "490adb0aa0e73f1372694cc2948e1692119e2771ae808a7a2fa65638811d28e3.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "DDB_TABLE": "hoge-table",
          },
        },
        "Handler": "index.putItem",
        "Layers": [
          "arn:aws:lambda:ap-northeast-1:580247275435:layer:LambdaInsightsExtension:14",
        ],
        "MemorySize": 256,
        "Role": {
          "Fn::GetAtt": [
            "LambdaConstructputItemServiceRole1DA96156",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 30,
        "TracingConfig": {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "LambdaConstructputItemLogRetention9DE23315": {
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/lambda/",
              {
                "Ref": "LambdaConstructputItem09D64D38",
              },
            ],
          ],
        },
        "RetentionInDays": 90,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LambdaConstructputItemServiceRole1DA96156": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/CloudWatchLambdaInsightsExecutionRolePolicy",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LambdaConstructputItemServiceRoleDefaultPolicy730237CB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LambdaConstructputItemServiceRoleDefaultPolicy730237CB",
        "Roles": [
          {
            "Ref": "LambdaConstructputItemServiceRole1DA96156",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": "cdk-hnb659fds-assets-729929860662-ap-northeast-1",
          "S3Key": "5fa1330271b8967d9254ba2d4a07144f8acefe8b77e6d6bba38261373a50d5f8.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
